examples/basic.rs

```diff
@@ -30,51 +23,48 @@ fn test_data() -> TestData {
 /// Creates a connected pair of sockets.
 ///
 /// This is similar to UnixStream::socket_pair, but works on windows too.
-///
-/// This is blocking, so it arguably doesn't belong into an async application, but this is not the
-/// point of the example here.
-fn socket_pair() -> Result<(TcpStream, TcpStream), Error> {
+async fn socket_pair() -> Result<(TcpStream, TcpStream), Error> {
     // port 0 = let the OS choose
-    let listener = StdTcpListener::bind("127.0.0.1:0")?;
-    let stream1 = StdTcpStream::connect(listener.local_addr()?)?;
-    let stream2 = listener.accept()?.0;
-    let stream1 = TcpStream::from_std(stream1, &Handle::default())?;
-    let stream2 = TcpStream::from_std(stream2, &Handle::default())?;
+    let mut listener = TcpListener::bind("127.0.0.1:0").await?;
+    let stream1 = TcpStream::connect(listener.local_addr()?).await?;
+    let stream2 = listener.accept().await?.0;
+
     Ok((stream1, stream2))
 }
 
-fn main() -> Result<(), Error> {
+#[tokio::main]
+async fn main() -> Result<(), Error> {
     // This creates a pair of TCP domain sockets that are connected together.
-    let (sender_socket, receiver_socket) = socket_pair()?;
+    let (sender_socket, receiver_socket) = socket_pair().await?;
 
     // Create the codec, type annotations are needed here.
     let codec: Codec<TestData, TestData> = Codec::new();
 
     // Get read and write parts of our streams (we ignore the other directions, but we could
     // .split() them if we wanted to talk both ways).
-    let sender = codec.clone().framed(sender_socket);
+    let mut sender = codec.clone().framed(sender_socket);
     let receiver = codec.framed(receiver_socket);
 
     // This is the data we will send over.
     let msg1 = test_data();
     let msg2 = test_data();
 
+    let mut msgs = futures::stream::iter(vec![Ok(msg1), Ok(msg2)]);
+
     // Send method comes from Sink and it will return a future we can spawn with tokio.
-    // It consumes self, so we need to chain the next send with then.
-    let send_all = sender
-            .send(msg1)
-            .and_then(|sender| sender.send(msg2))
-            // Close the sink (otherwise it would get returned throughout the join and block_on and
-            // the for_each would wait for more messages).
-            .map(|sender| drop(sender));
+    sender.send_all(&mut msgs).await?;
+
+    // Close the sink (otherwise it would get returned throughout the join and block_on and
+    // the for_each would wait for more messages).
+    sender.close().await?;
```

Comment:
Technically, this might deadlock. This part tries to send everything, but nothing currently reads. I guess it mostly works because OS buffers are large enough, but strictly speaking this is probably wrong.

The original had the join, so they run in parallel ‒ but this code is sequential and insists on sending everything first and only then receiving.

I don't know if you want to do joins or spawn the sending part, but it would be nice to show it „properly“ in the example.

Reply:
I've updated the example code to wrap sending and closing into one future and added some comments.